### ISSUE TITLE
[AArch64] BOLT asserts when attempting to encode out-of-range Pending Relocations

### ISSUE THREAD

**Issue by paschalis-mpeis** [llvm/llvm-project#116817]
## What is the problem

When a function is ignored from BOLT optimization, some relocations are emitted to patch the original fixups in the original text. Those are generated during external reference scanning. This is an optimization for adjusting the initial calls to land back to the optimized code. The relevant relocations are of type `R_AARCH64_CALL26`.
**In large binaries the encoding does not fit the supported range, leading to an [assertion](https://github.com/llvm/llvm-project/blob/67e733d36ef0c5ec9fab899ecf9f191d83c7d418/bolt/lib/Core/Relocation.cpp#L387):**
> only PC +/- 128MB is allowed for direct call

Ignoring functions can be common. Some reasons include:
- they are explicitly requested in skip-funcs (like in the below reproducer)
- could be a cold function in split mode
- or could be implicitly ignored for other reasons

---

# Reproducer

### How the reproducer triggers the crash
The error is triggered through instrumentation due to code size increases. 

When instrumenting MongoDB, BOLT requires to ignore some functions in order to succeed. When doing so, some additional relocations are emitted. However, instrumentation ends up increasing code size by more than 3x, which goes beyond those relocations' range. This causes an assertion crash when trying to encode such relocations (`R_AARCH64_CALL26`).

### Why BOLT requires to ignore some functions
This requirement was introduced by PR [#89681](https://github.com/llvm/llvm-project/pull/89681). Before that stricter check, those functions were not required to be ignored and the bug was not triggered (ie, on LLVM 18 RC). PR [#101466](https://github.com/llvm/llvm-project/pull/101466) improves support for entry-points other than the main one to the function, but still BOLT bails out in our use case. 

Regardless, this issue will not focus on such improvements as the error can be triggered without it

### Reproducer Instructions
- Tested on: Ubuntu 22.04.5 LTS, (6.8.0-1018-aws)

Use use this [Dockerfile](https://gist.githubusercontent.com/paschalis-mpeis/9eb878f73e18fb9d3f996ae7c59d4792/raw/5baef8dc9acdd1cc74800aa5d3542993c926129d/Dockerfile) and the below script to compile the input binary and pull it on an AArch64 Ubuntu host.
Alternatively, one can get it pre-compiled from [here](https://gist.github.com/paschalis-mpeis/9eb878f73e18fb9d3f996ae7c59d4792/raw/e9de7728a517e153d4980225ecac92d2e0923498/mongod.tar.gz).

```bash
# Compile mongodb v7.0.5 with clang 18.1.8 (stage-0):
docker build --progress=plain --tag 'tmp-mdb-stage-0' .

# Then, 'pull' the binary from container to the host, given host is Ubuntu 22.04/24.04 (preferred)
docker run --rm --entrypoint cat tmp-mdb-stage-0 mongo/build/install/bin/mongod > mongod
chmod +x mongod

# once binary is retrieved the image could be deleted
docker rmi tmp-mdb-stage-0
```

Then, compile bolt version `dcd62070cf45` **with assertions ON** and use it to instrument the binary with:
```bash
LLVM_DIR=path/to/llvm/bin/dir

$LLVM_DIR/llvm-bolt mongod -instrument -o mongod.instrumented \
	--instrumentation-file=prof.fdata \
	--instrumentation-sleep-time=60 \
	--instrumentation-no-counters-clear \
	--instrumentation-wait-forks \
	--skip-funcs=_ZL9InterpretP9JSContextRN2js8RunStateE/1,_ZN2v88internal12_GLOBAL__N_18RawMatchIhEENS0_19IrregexpInterpreter6ResultEPNS0_7IsolateENS0_9ByteArrayENS0_6StringENS0_6VectorIKT_EEPiiiijNS0_6RegExp10CallOriginEj/1,_ZN2v88internal12_GLOBAL__N_18RawMatchIDsEENS0_19IrregexpInterpreter6ResultEPNS0_7IsolateENS0_9ByteArrayENS0_6StringENS0_6VectorIKT_EEPiiiijNS0_6RegExp10CallOriginEj/1
```

---
## Proposed Solutions
A draft PR around (S1) will follow.

#### (S1) Apply pending relocs only when in bounds
Keep the 'external reference scan' optimization on, and apply it on a per callsite basis.
That is, if a particular external reference is within range, then the encoding of that relocation happens, otherwise it is ignored.

This may require some checks that the PendingRelocations were added specifically for optimization, ie during external reference scanning and that are `R_AARCH64_CALL26`.

#### (S2) Hint users to use `-no-scan`
Instead of the [current assertion](https://github.com/llvm/llvm-project/blob/67e733d36ef0c5ec9fab899ecf9f191d83c7d418/bolt/lib/Core/Relocation.cpp#L387), BOLT could escalate this to an exit with a recommendation to use `-no-scan` to overcome this crash. This is not very obvious unless someone is familiar with specific BOLT code regions. Also, by bailing out at all times, we ensure that BOLT does not patch the code in some wrong way?

Q: What if an 'external reference' points to a cold function? Is that pending relocation still flushed?
I assume that even if it does, it will have no effect.

**Comment by maksfb** [comment]
S1 sounds good to me. We just have to make sure that patching of old code is enabled, e.g. via `opts::ForcePatch`.

> Q: What if an 'external reference' points to a cold function? Is that pending relocation still flushed?
I assume that even if it does, it will have no effect.

If a function is not going to be emitted, then there will be no reference/relocation recorded. Cold functions are not emitted in "lite" mode which is default for x86, but not AArch64.

https://github.com/llvm/llvm-project/blob/14a259f85b6cbe6827677d94990c8803e31c847d/bolt/lib/Core/BinaryFunction.cpp#L1565-L1576

**Comment by paschalis-mpeis** [comment]
> S1 sounds good to me. We just have to make sure that patching of old code is enabled, e.g. via `opts::ForcePatch`.

I may be missing some connection between [scanExternalRefs](https://github.com/llvm/llvm-project/blob/1ab81f8e7f77110c4a752dd7d2cc39fb5148760c/bolt/lib/Core/BinaryFunction.cpp#L1507)+[flushPendingRelocations](https://github.com/llvm/llvm-project/blob/1ab81f8e7f77110c4a752dd7d2cc39fb5148760c/bolt/lib/Core/BinarySection.cpp#L154) and [PendingPatches](https://github.com/llvm/llvm-project/blob/1ab81f8e7f77110c4a752dd7d2cc39fb5148760c/bolt/lib/Passes/PatchEntries.cpp#L74) of PatchEntries.cpp.

Do you want to introduce another requirement for when scanExternalRefs/flushPendingRelocations should run? ie only when `opts:ForcePatch` was also set? (on top of requiring [no-scan](https://github.com/llvm/llvm-project/blob/76f258920d3baf32be297f60bee5b8520f195c25/bolt/lib/Core/BinaryFunction.cpp#L1515) not to be set)

**Comment by maksfb** [comment]
> > S1 sounds good to me. We just have to make sure that patching of old code is enabled, e.g. via `opts::ForcePatch`.
> 
> I may be missing some connection between [scanExternalRefs](https://github.com/llvm/llvm-project/blob/1ab81f8e7f77110c4a752dd7d2cc39fb5148760c/bolt/lib/Core/BinaryFunction.cpp#L1507)+[flushPendingRelocations](https://github.com/llvm/llvm-project/blob/1ab81f8e7f77110c4a752dd7d2cc39fb5148760c/bolt/lib/Core/BinarySection.cpp#L154) and [PendingPatches](https://github.com/llvm/llvm-project/blob/1ab81f8e7f77110c4a752dd7d2cc39fb5148760c/bolt/lib/Passes/PatchEntries.cpp#L74) of PatchEntries.cpp.
> 
> Do you want to introduce another requirement for when scanExternalRefs/flushPendingRelocations should run? ie only when `opts:ForcePatch` was also set? (on top of requiring [no-scan](https://github.com/llvm/llvm-project/blob/76f258920d3baf32be297f60bee5b8520f195c25/bolt/lib/Core/BinaryFunction.cpp#L1515) not to be set)

Let me add more context on why we need to patch the original function entries. Whenever we duplicate a function body for execution purposes, we also need to duplicate all associate metadata, e.g. .eh_frame info, exception ranges, DWARF debug info, etc. If the function uses jump tables, those need to be duplicated as well. Currently, BOLT cannot duplicate all of the accompanying metadata and instead we force the execution of the new/optimized version while prohibiting execution of the original code. This is achieved in `PatchEntries` pass (https://github.com/llvm/llvm-project/blob/b35749559ddd9b2d4e044ef71d13d888b8a3d8cb/bolt/lib/Passes/PatchEntries.cpp#L34).

For the reasons described above, it makes sense to patch entries in all original functions in most of the cases. Since we care about the performance, we have an exception for "small" functions (https://github.com/llvm/llvm-project/blob/b35749559ddd9b2d4e044ef71d13d888b8a3d8cb/bolt/lib/Passes/PatchEntries.cpp#L66-L69). I plan to make the requirements for this exception case more strict.

There are also cases where the code cannot be patched due to size limitations. Whenever we fail to patch the original entries, we have to bail out on optimizing the function (https://github.com/llvm/llvm-project/blob/b35749559ddd9b2d4e044ef71d13d888b8a3d8cb/bolt/lib/Passes/PatchEntries.cpp#L116). Such cases are quite rare and we report them to the user.

Lastly, if we can guarantee that the old code never gets executed, we can skip patching altogether. This guarantee is achieved via successful run of `scanExternalRefs()` and the relocation processing that follows.

If we are going to bail out on relocation processing for `scanExternalRefs()`, we can no longer skip the patching process as the assumption that the old code is never executed becomes broken.

I'm going to experiment with enabling unconditional execution of `PatchEntries` on AArch64 to make "lite" mode functional.

**Comment by paschalis-mpeis** [comment]
Hey Maks,

Thanks for the detailed response, this is clearer now:
- `scanExternalRefs` ensures old code never executes
- If it fails, `PatchEntries` must run to enforce execution of only new code
- So both ensure the same effect regarding execution of old/new code

As a prerequisite to my patch, I opened [PR #127812](https://github.com/llvm/llvm-project/issues/116817#) to move `-force-patch` to bolt-core. I also summarized your explanation in PatchEntries header. If these are AArch64-specific limitations, I can note that.

I will then adjust my patch to enforce your suggestion.

---

Currently, if a user passes `-no-scan` without `-force-patch`, we’d be in the situation you describe. Should we exit or warn in this case?

> I'm going to experiment with enabling unconditional execution of PatchEntries on AArch64 to make "lite" mode functional.

Great. Do you expect performance benefits? Lite mode currently regresses, even with instrumentation, so profile quality isn’t the issue.

Also, do the listed limitations relate to the [skip-funcs AArch64 limitations](https://github.com/llvm/llvm-project/issues/126000#issuecomment-2638810314)? (if that is a quick answer)

**Comment by maksfb** [comment]
Another functionality of `scanExternalRefs()` is the update of function pointer references in cold and non-simple functions. If we don't create CFG for a function for any reason, we *must* update references to optimized functions, otherwise function pointer comparison will not work properly. I'm making this work for AArch64.

**Comment by maksfb** [comment]
>  Lite mode currently regresses, even with instrumentation, so profile quality isn’t the issue.

This might be a fallout from the current limitations of `scanExternalRefs()` on AArch64.

**Comment by paschalis-mpeis** [comment]
Oh, right, thank you. So we might end up staying longer in the old code it seems.
In the meantime I've updated the PR to satisfy the requirement
- #116964


(I'm not so confident if it'll further contribute to those issues though)

**Comment by paschalis-mpeis** [comment]
Solved by:
- #116964
